api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -267,10 +270,18 @@ public static Cohort union(Cohort a, Cohort b) {
 	public static Cohort intersect(Cohort a, Cohort b) {
 		Cohort ret = new Cohort();
 		ret.setName("(" + (a == null ? "NULL" : a.getName()) + " * " + (b == null ? "NULL" : b.getName()) + ")");
-		if (a != null && b != null) {
-			ret.getMemberships().addAll(a.getMemberships());
-			ret.getMemberships().retainAll(b.getMemberships());
+		List<CohortMembership> cohortMemberships = new ArrayList<CohortMembership>();
+			
+		for(CohortMembership  a_member: a.getMemberships()){
+			for(CohortMembership  b_member: b.getMemberships()){
+				if(a_member.getPatientId() == b_member.getPatientId()){
+					if(a_member.getStartDate().before(b_member.getStartDate())){
+						cohortMemberships.add(b_member);
```

Comment:
Is there any way to check this without two loops? If each one has `x `memberships, then the loop want to  go for `x*x` times to find it.

Reply:
Sure, working on this

Reply:
@suthagar23 I have tried removing the nested for loops but am hitting a snag. Would you be able to offer up a solution?